protos/encodings.proto

```diff
@@ -77,28 +77,33 @@ message Nullable {
 message List {
     // An array containing the offsets into an items array.
     //
-    // This array will have (num_rows + 1), will have a data
-    // type of uint64, and will never have nulls.
+    // This array will have num_rows items and will never
+    // have nulls.
     //
-    // offsets[0] will always be 0.
-    //
-    // If the incoming list at index i is not null then offsets[i+1]
-    // will contain offsets[i] + len(list)
+    // If the list at index i is not null then offsets[i] will
+    // contain `base + len(list)` where `base` is defined as:
+    //   i == 0: 0
+    //   i >  0: (offsets[i-1] % first_invalid_offset)
```

Comment:
> Compute the sizes of each array ([2, 0, 1, 0, 3])
> Transform the sizes to cumulative sum ([2, 2, 3, 3, 7])

Yes.  If there are no nulls then this is the exact same thing as Arrow's list encoding except the leading 0 has been removed (as it is redundant, though I could be convinced to add it back in, it's about the same amount of complexity either way though so I figured we might as well save a few bytes)

That being said, I think you want `[2, 2, 3, 3, 6]` as the last list has 3 items, not 4.  The arrow offsets will be `[0, 2, 2, 3, 3, 6]`.

> Choose a value for first_invalid_offset that's greater than any existing offset (e.g. 100)

No.  You can't pick any invalid value.  It needs to be the last valid offset + 1.  In other words, given the example of `[2, 2, 3, 3, 7]` it must be `8`.

> Add the first_invalid_offset to any slot that is null ([2, 2, 3, 103, 7])

Yes.

> And then on the read size, seems like we can do:

Yes, more or less.

> Otherwise, the range for the list is 0, offsets[0].

Not if `offsets[0] >= first_invalid_offset` which will happen if the first list is null.

> Otherwise, the range is offsets[i - 1] % first_invalid_offset, offsets[i]

It's not exactly `%`.  I thought it was originally, but I was wrong.  If we used `7` as `first_invalid_offset` then it would be `%` but unfortunately, we can't do that, because then we can't tell if the last item in the list is null or not.  Thanks for pointing this out.  I will clean this up.  It should be...

```
if offsets[i - 1] >= first_invalid_offset { offsets[i - 1] - first_invalid_offset - 1 } else { offsets[i - 1] }, offsets[i]
```

Reply:
> It would be nice to have an example.

I will add one.  There is an example in the decoder that I walk through but I agree we need one here too.

Reply:
Ok.  The reason I was thinking it needed to be **first** (and not any arbitrary invalid offset) is because we sometimes want to know the number of items referenced by the list before we've done any actually loading of offsets (for scheduling purposes).  I think, however, there are other ways the scheduler could get this information (e.g. pages have a num_rows and so the list scheduler could add up the num_rows on all the assigned item pages) so this may not be a valid requirement any more.

However, from a performance perspective, there is another reason we want to use the first invalid offset.  By using this we force the range of encoded values to be `[0, 2N]` which means we are much more likely to get good results from bit packing.

Also, I see that in some places I call it the `last_valid_offset` and in others I call it the `first_invalid_offset` :facepalm: .  I will clean this up as well.

Reply:
And I can see how "first valid offset" might be construed as "first offset where a list item is valid (not null)".  Maybe I will call it "null offset adjustment" instead

Reply:
At this point it's probably easiest to ignore everything I said in this entire thread :laughing: I've added an example to the .proto file so start there instead.